File: src/services/userService.js
// 사용자 서비스 — 로그인/조회 도메인 로직 (HTTP 비의존, ADR-006). 의존성은 주입(userModel).
// 비밀번호(평문/해시)는 어떤 반환값에도 포함하지 않는다(news.md 보안).
// 로그인은 사용자 존재 여부와 무관하게 항상 bcrypt 비교를 1회 수행해 타이밍 차이를 줄인다.

import bcrypt from 'bcryptjs';

// 응답에 노출 가능한 사용자 필드 — password 및 잠금 상태(failedLoginCount/lockedUntil/
// lastFailedLoginAt)는 의도적으로 제외한다(계정 열거 단서 차단).
const SAFE_FIELDS = ['userId', 'name', 'role', 'department', 'departmentCode', 'active'];

// 계정 잠금 기본 정책 — IP 단위 레이트리밋과 보완 관계인 사용자 단위 누적 실패 잠금.
const DEFAULT_LOCKOUT_THRESHOLD = 5;
const DEFAULT_LOCKOUT_WINDOW_MS = 15 * 60 * 1000;

// 사용자가 없을 때도 동일한 비용의 bcrypt 비교를 수행하기 위한 더미 해시(타이밍 공격 완화).
const DUMMY_HASH = bcrypt.hashSync('*timing-equalizer*', 10);

function sanitize(row = {}) {
  const out = {};
  for (const f of SAFE_FIELDS) if (row[f] !== undefined) out[f] = row[f];
  return out;
}

export function createUserService({
  userModel,
  now = () => Date.now(),
  lockoutThreshold = DEFAULT_LOCKOUT_THRESHOLD,
  lockoutWindowMs = DEFAULT_LOCKOUT_WINDOW_MS,
}) {
  // 아이디/비밀번호 대조 로그인. 비밀번호는 반환하지 않는다.
  // 존재하지 않는 사용자도 더미 해시로 비교해 성공/실패 경로의 소요 시간 차이를 최소화한다.
  // 잠긴 계정도 비밀번호 비교를 1회 수행해 잠금/비잠금 간 소요 시간 차이를 만들지 않는다.
  async function login(userId, password) {
    const row = userModel.findById(userId);
    const hash = (row && row.password) || DUMMY_HASH;
    // 잠긴 경우에도 bcrypt 비교를 1회 수행해 성공/실패/잠금 경로의 소요 시간 차이를 줄인다.
    const passwordOk = await bcrypt.compare(String(password ?? ''), hash);

    // 비활성 거부는 잠금 판정·카운트보다 우선한다(비활성 계정에 카운트를 올리지 않는다).
    if (row && row.active === 'N') return { ok: false, reason: 'inactive' };

    // 자격이 틀리면 실패 카운트를 누적하고 임계치 도달 시 잠근다.
    if (!row || !passwordOk) {
      if (row) registerFailure(row);
      return { ok: false, reason: 'invalid-credentials' };
    }

    // 잠금 만료는 lockedUntil을 주입 시계와 비교해 판정한다(행 삭제 없음).
    if (isLocked(row)) return { ok: false, reason: 'locked' };

    // 성공 — 잠금 카운트/잠금 시각을 리셋해 영속한다.
    resetLockout(row);
    return { ok: true, user: sanitize(row) };
  }

  // lockedUntil이 현재 시각보다 미래이면 잠긴 것으로 본다.
  function isLocked(row) {
    const until = Number(row.lockedUntil);
    return Number.isFinite(until) && until > now();
  }

  // 실패 1회 누적 — 임계치 이상이면 lockedUntil을 설정한다.
  function registerFailure(row) {
    const next = (Number(row.failedLoginCount) || 0) + 1;
    const patch = { failedLoginCount: String(next), lastFailedLoginAt: String(now()) };
    if (next >= lockoutThreshold) patch.lockedUntil = String(now() + lockoutWindowMs);
    userModel.update(row.userId, patch);
  }

  // 성공/만료 시 카운트·잠금 상태를 초기화한다(비파괴: 행은 유지, 필드만 비움).
  function resetLockout(row) {
    userModel.update(row.userId, {
      failedLoginCount: '0', lockedUntil: '', lastFailedLoginAt: '',
    });
  }

  // 정제된 사용자 목록(비밀번호 제외).
  function query(filters) {
    return userModel.query(filters).map(sanitize);
  }

  // 사용자 생성 — 비밀번호는 bcrypt 해시로 저장하고, 반환값엔 비밀번호를 담지 않는다.
  function create(dto = {}) {
    const { password, ...rest } = dto;
    const row = {
      ...rest,
      password: bcrypt.hashSync(String(password ?? ''), 10),
      active: rest.active ?? 'Y',
    };
    userModel.insert(row);
    return { ok: true, user: sanitize(row) };
  }

  // 사용자 수정 — 비밀번호가 오면 해시로 저장한다(없으면 그대로 둠). 비밀번호는 반환하지 않는다.
  function update(userId, fields = {}) {
    const patch = { ...fields };
    if (patch.password !== undefined) patch.password = bcrypt.hashSync(String(patch.password), 10);
    const changes = userModel.update(userId, patch);
    return { ok: true, changes };
  }

  return { login, query, create, update };
}

File: server-spring/src/main/java/harness/server/UserService.java
package harness.server;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.function.LongSupplier;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;

/**
 * 사용자 도메인 서비스 — 로그인/잠금 로직(HTTP 비의존, ADR-006). userService.js 1:1 이식.
 *
 * <p>규율:
 * <ul>
 *   <li>존재하지 않는 사용자도 더미 해시로 bcrypt 비교 1회 → 성공/실패 경로 타이밍 차이 완화(계정 열거 방지).</li>
 *   <li>비활성(active='N') 거부는 잠금 판정·카운트보다 <b>우선</b>(비활성 계정에 카운트 올리지 않음).</li>
 *   <li>잠긴 계정은 <b>올바른 비밀번호여도</b> locked(자격 판정보다 잠금 우선).</li>
 *   <li>실패 누적은 임계(기본 5) 도달 시 lockedUntil=now+windowMs 설정, 성공 시 리셋(행 삭제 없음 — 필드만).</li>
 *   <li>비밀번호(평문/해시)는 어떤 반환값에도 없다 — 반환은 {@link UserRow}(상위가 Projections 로 투영).</li>
 * </ul>
 * bcrypt 검증은 spring-security-crypto {@link BCryptPasswordEncoder}(bcryptjs $2a$ 해시 호환).
 */
public class UserService {

    public static final int LOCKOUT_THRESHOLD = 5;
    public static final long LOCKOUT_DURATION_MS = 15L * 60L * 1000L;

    private final BCryptPasswordEncoder bcrypt = new BCryptPasswordEncoder();
    // 사용자가 없을 때도 동일 비용의 bcrypt 비교를 하기 위한 더미 해시(타이밍 완화).
    private final String dummyHash = bcrypt.encode("*timing-equalizer*");

    private final UserRepository users;
    private final LongSupplier now;
    private final int threshold;
    private final long windowMs;

    public UserService(UserRepository users, LongSupplier now) {
        this(users, now, LOCKOUT_THRESHOLD, LOCKOUT_DURATION_MS);
    }

    public UserService(UserRepository users, LongSupplier now, int threshold, long windowMs) {
        this.users = users;
        this.now = now;
        this.threshold = threshold;
        this.windowMs = windowMs;
    }

    /** 로그인 결과 — 성공 시 row(안전 투영은 상위가 한다), 실패 시 reason 토큰. */
    public record LoginResult(boolean ok, String reason, UserRow row) {
        static LoginResult fail(String reason) {
            return new LoginResult(false, reason, null);
        }

        static LoginResult success(UserRow row) {
            return new LoginResult(true, null, row);
        }
    }

    public LoginResult login(String userId, String password) {
        UserRow row = (userId == null || userId.isEmpty()) ? null : users.findUser(userId);
        String hash = (row != null && row.password() != null) ? row.password() : dummyHash;
        // 잠긴/없는 경우에도 bcrypt 비교를 1회 수행(경로 간 소요 시간 차이 완화).
        boolean passwordOk = bcrypt.matches(password == null ? "" : password, hash);

        // 비활성 거부는 잠금·카운트보다 우선.
        if (row != null && "N".equals(row.active())) {
            return LoginResult.fail("inactive");
        }
        // 잠긴 계정은 올바른 비밀번호여도 거부(자격보다 잠금 우선).
        if (row != null && isLocked(row)) {
            return LoginResult.fail("locked");
        }
        // 자격 불일치 — 존재하는 계정이면 실패 누적. 응답은 존재/부재 구분 없이 동일.
        if (row == null || !passwordOk) {
            if (row != null) {
                registerFailure(row);
            }
            return LoginResult.fail("invalid-credentials");
        }
        // 성공 — 잠금 카운트/시각 리셋.
        resetLockout(row);
        return LoginResult.success(row);
    }

    private boolean isLocked(UserRow row) {
        Long until = parseLong(row.lockedUntil());
        return until != null && until > now.getAsLong();
    }

    private void registerFailure(UserRow row) {
        long current = parseLongOrZero(row.failedLoginCount());
        long next = current + 1;
        Map<String, Object> patch = new LinkedHashMap<>();
        patch.put("failedLoginCount", String.valueOf(next));
        patch.put("lastFailedLoginAt", String.valueOf(now.getAsLong()));
        if (next >= threshold) {
            patch.put("lockedUntil", String.valueOf(now.getAsLong() + windowMs));
        }
        users.updateUser(row.userId(), patch);
    }

    private void resetLockout(UserRow row) {
        Map<String, Object> patch = new LinkedHashMap<>();
        patch.put("failedLoginCount", "0");
        patch.put("lockedUntil", null); // SQL NULL — "잠금 없음"이 명확.
        patch.put("lastFailedLoginAt", null);
        users.updateUser(row.userId(), patch);
    }

    private static Long parseLong(String s) {
        if (s == null || s.isEmpty()) {
            return null;
        }
        try {
            return Long.valueOf(s);
        } catch (NumberFormatException e) {
            return null;
        }
    }

    private static long parseLongOrZero(String s) {
        Long v = parseLong(s);
        return v == null ? 0L : v;
    }
}
